pkg/executor/join/hash_join_v2.go

```diff
@@ -677,166 +629,158 @@ func (e *HashJoinV2Exec) createTasks(buildTaskCh chan<- *buildTask, totalSegment
 				startIndex := partitionStartIndex[partIdx]
 				endIndex := min(startIndex+segStep, partitionSegmentLength[partIdx])
 				select {
-				case <-doneCh:
-					return
+				case <-syncer:
+					// syncer is closed by build fetcher in advance,
+					// this means that there happen some errors.
+					return false, nil
 				case buildTaskCh <- createBuildTask(partIdx, startIndex, endIndex):
 				}
 				partitionStartIndex[partIdx] = endIndex
 				hasNewTask = true
 			}
+			continue
 		}
 		if !hasNewTask {
 			break
 		}
 	}
+
+	return true, nil
 }
 
-func (e *HashJoinV2Exec) fetchAndBuildHashTable(ctx context.Context) {
-	if e.stats != nil {
-		start := time.Now()
-		defer func() {
-			e.stats.fetchAndBuildHashTable = time.Since(start)
-		}()
-	}
+func (e *HashJoinV2Exec) dispatchBuildTasks(syncer chan struct{}) error {
+	defer close(e.buildFinished)
+	ifContinue := true
+	var err error
 
-	waitJobDone := func(wg *sync.WaitGroup, errCh chan error) bool {
-		wg.Wait()
-		close(errCh)
-		if err := <-errCh; err != nil {
-			e.buildFinished <- err
-			return false
+	for ifContinue {
+		ifContinue, err = e.dispatchBuildTasksImpl(syncer)
```

Comment:
dispatchBuildTasksImpl can be called multiple times here?

Reply:
> dispatchBuildTasksImpl can be called multiple times here?

Yes, after spill is added, this function will be called multiple times/